src/sources/fluent.rs

```diff
@@ -0,0 +1,1040 @@
+        match message {
+            FluentMessage::Message(tag, timestamp, record)
+            | FluentMessage::MessageWithOptions(tag, timestamp, record, ..) => {
+                self.unread_frames.push_back(FluentFrame {
+                    tag: tag.clone(),
+                    timestamp,
+                    record,
+                });
+                Ok(())
+            }
+            FluentMessage::Forward(tag, entries)
+            | FluentMessage::ForwardWithOptions(tag, entries, ..) => {
+                self.unread_frames.extend(entries.into_iter().map(
+                    |FluentEntry(timestamp, record)| FluentFrame {
+                        tag: tag.clone(),
+                        timestamp,
+                        record,
+                    },
+                ));
+                Ok(())
+            }
+            FluentMessage::PackedForward(tag, bin) => {
+                let mut buf = BytesMut::from(&bin[..]);
+
+                let mut decoder = FluentEntryStreamDecoder;
+
+                while let Some(FluentEntry(timestamp, record)) = decoder.decode(&mut buf)? {
+                    self.unread_frames.push_back(FluentFrame {
+                        tag: tag.clone(),
+                        timestamp,
+                        record,
+                    });
+                }
+                Ok(())
+            }
+            FluentMessage::PackedForwardWithOptions(tag, bin, options) => {
+                let buf = match options.compressed.as_deref() {
+                    Some("gzip") => {
+                        let mut buf = Vec::new();
+                        MultiGzDecoder::new(io::Cursor::new(bin.into_vec()))
+                            .read_to_end(&mut buf)
+                            .map(|_| buf)
+                            .map_err(Into::into)
+                    }
+                    Some("text") | None => Ok(bin.into_vec()),
+                    Some(s) => Err(DecodeError::UnknownCompression(s.to_owned())),
+                }?;
+
+                let mut buf = BytesMut::from(&buf[..]);
+
+                let mut decoder = FluentEntryStreamDecoder;
+
+                while let Some(FluentEntry(timestamp, record)) = decoder.decode(&mut buf)? {
+                    self.unread_frames.push_back(FluentFrame {
+                        tag: tag.clone(),
+                        timestamp,
+                        record,
+                    });
+                }
+                Ok(())
+            }
+            FluentMessage::Heartbeat(rmpv::Value::Nil) => Ok(()),
+            FluentMessage::Heartbeat(value) => Err(DecodeError::UnexpectedValue(value)),
+        }
+    }
+}
+
+impl Decoder for FluentDecoder {
+    type Item = FluentFrame;
+    type Error = DecodeError;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        if !self.unread_frames.is_empty() {
+            return Ok(self.unread_frames.pop_front());
+        }
```

Comment:
Don't need to check `is_empty` first:
```suggestion
        if let Some(frame) = self.unread_frames.pop_front() {
            return Ok(frame);
        }
```